### ISSUE TITLE
History no longer shown for renamed files

### ISSUE THREAD

**Issue by CharonX2003** [gitextensions/gitextensions#5944]
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
History is not shown for files before their rename (but blame correctly identifies commits)

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**
Take a file with history, rename it, perform additional commits after changing it.

**What is the expected behavior?**
History is shown for the file before it got renamed (as it currently works for Blame)

**Environment you encounter the issue:**
- Git Extensions 3.00.00.4433
- Build fca7cf228b481ee8c1b779cf7b882ccdfbdcd1bc
- Git 2.20.0.windows.1
- Microsoft Windows NT 10.0.17134.0
- .NET Framework 4.7.3260.0
- DPI 96dpi (no scaling)

(In settings "Follow renames in History" is enabled, "Follow exact renames and copies only" has been tried as both disabled and enabled - the files have 100% similarity".)

**Did this work in previous version of GitExtensions (which)?**

I recall the history being shown correctly for the renamed files in Version 2.5x

**Comment by vbjay** [comment]
Notice rename settings
![image](https://user-images.githubusercontent.com/3629489/50184226-6dd43900-02e2-11e9-81b9-ee87ca4ae2e7.png)

**Comment by mstv** [comment]
I have never seen it working. Renaming/cloning files is Git's great weakness.
Here is a sample branch with renamed `appveyor.yml`:
https://github.com/mstv/gitextensions/tree/repro/nre_getbrushforrevision

**Comment by gerhardol** [comment]
Works for me

**Comment by mstv** [comment]
@gerhardol, do you really see commits from before the file was renamed?!?

**Comment by drewnoakes** [comment]
@mstv what git arguments do you see in the command log for this?

**Comment by drewnoakes** [comment]
@CharonX2003 you could also check the command log and let us know what git arguments are being used when you see the problem.

**Comment by mstv** [comment]
`check-attr -z diff text crlf eol -- "appveyorX.yml"`

**Comment by mstv** [comment]
and
`-c log.showSignature=false log -z --pretty=format:"%H%T%P%n%at%n%ct%n%e%n%aN%n%aE%n%cN%n%cE%n%s%n%n%b" --all --boundary --max-count=100000 --name-only --parents --find-renames --find-copies -- "appveyorX.yml" "fatal: unrecognized argument: --format"`
regardless whether `Show Full History` ![grafik](https://user-images.githubusercontent.com/36601201/50291686-84f46180-046f-11e9-88dc-4d0eaec09aeb.png) is on or off
I doubt that `"fatal: unrecognized argument: --format"` is a correct git argument.
This should have thrown an exception instead in my opinion.

**Comment by CharonX2003** [comment]
I'm also seeing the "fatal: unrecognized argument: --format" string...

File name:   C:\Program Files\Git\bin\git.exe
Arguments:   -c log.showSignature=false log --format="%n" --name-only --format  --find-renames --find-copies -- "src/main.cpp"

File name:   C:\Program Files\Git\bin\git.exe
Arguments:   show-ref --dereference

File name:   C:\Program Files\Git\bin\git.exe
Arguments:   -c log.showSignature=false log -z --pretty=format:"%H%T%P%n%at%n%ct%n%e%n%aN%n%aE%n%cN%n%cE%n%s%n%n%b" --all --boundary --not --glob=notes --not --max-count=100000 --full-history --simplify-merges  --name-only --parents --find-renames --find-copies -- "src/main.cpp" "fatal: unrecognized argument: --format"

File name:   C:\Program Files\Git\bin\git.exe
Arguments:   -c diff.submodule=short -c diff.noprefix=false diff --no-color --ignore-all-space --unified=3 -M -C "14b663c94e0d959622f2d0cb8b5d40875290d58d" "266db14452f69cad00a8636db3d26300249c7f1c" -- "src/main.cpp"

**Comment by drewnoakes** [comment]
@mstv @CharonX2003 looks like you found the clue that'll break this.

The string `fatal: unrecognized argument: --format` looks like it came out of git, suggesting that an earlier command produced that string and we blindly passed it forward to the log command.

Is there an earlier command with a non-zero exit code (5th column)?

**Comment by mstv** [comment]
There is only:
```
21:25:17.652	104ms	27288		0	git	ls-files -z --unmerged	F:\Build\gitextensions\
21:25:23.051	101ms	19028		128	git	-c log.showSignature=false log --format="%n" --name-only --format  --find-renames --find-copies -- "appveyorX.yml"	F:\Build\gitextensions\
21:25:23.164	92ms	14068	UI	0	git	rev-parse HEAD	F:\Build\gitextensions\
21:25:23.259	119ms	21344	UI	0	git	show-ref --dereference	F:\Build\gitextensions\
21:25:23.385	251ms	8536		  	git	-c log.showSignature=false log -z --pretty=format:"%H%T%P%n%at%n%ct%n%e%n%aN%n%aE%n%cN%n%cE%n%s%n%n%b" --all --boundary --not --glob=notes --not --max-count=100000 --name-only --parents --find-renames --find-copies -- "appveyorX.yml" "fatal: unrecognized argument: --format"	F:\Build\gitextensions\
21:25:23.830	100ms	34268	UI	0	git	check-attr -z diff text crlf eol -- "appveyorX.yml"	F:\Build\gitextensions\
```

The output of `git	show-ref --dereference	F:\Build\gitextensions\` is empty (`\` replaced with `/`).

**Comment by drewnoakes** [comment]
It's the line with the `128`. That's git telling us it's upset. You can see the `--format` argument is present twice.

**Comment by mstv** [comment]
Yes, of course.

Does Git return non-zero results that often that GE avoids throwing an exception in this case?

**Comment by drewnoakes** [comment]
Ugh, apologies everyone. I accidentally pushed to this repo 00b8daa34e9d15fcb99f632900fd4fc6c9b9dcb1 because I was rushing. Rather than back it out and potentially cause more trouble, if anyone objects I will file a revert PR.

However hopefully that push fixes this issue. Sorry again.

**Comment by mstv** [comment]
The return code is 0 now. Though the number of displayed commits has not changed.
```
2018-12-20T21:45:37.3137927+01:00	85	36904		0	git	ls-files -z --unmerged	F:\Build\gitextensions\
2018-12-20T21:45:40.6227129+01:00	115	18088		0	git	ls-files -z --unmerged	F:\Build\gitextensions\
2018-12-20T21:45:42.9440076+01:00	158	15732		0	git	-c log.showSignature=false log --format="%n" --name-only  --find-renames --find-copies -- "appveyorX.yml"	F:\Build\gitextensions\
2018-12-20T21:45:43.1185298+01:00	79	14212	UI	0	git	rev-parse HEAD	F:\Build\gitextensions\
2018-12-20T21:45:43.2030405+01:00	102	13548	UI	0	git	show-ref --dereference	F:\Build\gitextensions\
2018-12-20T21:45:43.3465587+01:00	367	27344		0	git	-c log.showSignature=false log -z --pretty=format:"%H%T%P%n%at%n%ct%n%e%n%aN%n%aE%n%cN%n%cE%n%s%n%n%b" --all --boundary --max-count=100000 --full-history --simplify-merges  --name-only --parents --find-renames --find-copies -- "appveyorX.yml"	F:\Build\gitextensions\
2018-12-20T21:45:43.9601367+01:00	95	28968	UI	0	git	check-attr -z diff text crlf eol -- "appveyorX.yml"	F:\Build\gitextensions\
```

In general: Does Git return non-zero results that often that GE avoids throwing an exception in this case?

**Comment by RussKie** [comment]
you've misspelt it Drew, second "t" is missing 🤣 
![image](https://user-images.githubusercontent.com/4403806/50310241-7d779b80-04f5-11e9-808c-5f40cea0d439.png)

**Comment by drewnoakes** [comment]
> second "t" is missing

Haha. Yet another reason not to rush things....

> Does Git return non-zero results that often that GE avoids throwing an exception in this case?

Haven't looked into it. There's a single abstraction over process loading and execution, with different ways of reading out of such processes with different trade offs. It's possible that we can be more strict in at least some of those cases, if not all.

**Comment by drewnoakes** [comment]
> the number of displayed commits has not changed

I see the next problem, and it should be the last one. Let me try again, without rushing this time :)

**Comment by drewnoakes** [comment]
The code in `FormFileHistory` has some logic to try and work around a bug it references, with a now-dead URL about the `--follow` argument.

For the log command:

> -c log.showSignature=false log -z --pretty=format:"%H%T%P%n%at%n%ct%n%e%n%aN%n%aE%n%cN%n%cE%n%s%n%n%b" --all --boundary --max-count=100000 --full-history --simplify-merges  --name-only --parents --find-renames --find-copies -- "my-file.ext"

Adding `--follow` to the list of arguments produces a lot more output. However in #2363 a few years back it seems there was a problem with too much stuff appearing in that output.

This needs a more careful review than I have the time for right now as it's late here.

**Comment by RussKie** [comment]
@drauch what version are you running? 
@drewnoakes has pushed a fix into master about a month ago, have you tried the canary build?

**Comment by RussKie** [comment]
> this should be a showstopper bug :(

Whilst this bug can certainly be annoying and cause discomfort, I can't classify it as a "blocker" or a "showstopper", I am afraid. 
Showing history for renamed files isn't the primary or the only function of the app. 

We also appreciate help and accept pull requests with corrections.

**Comment by vbjay** [comment]
No to mention, git doesn't really track filenames per se.  This should explain it.  http://thisbythem.com/blog/preserving-history-when-renaming-files-in-git/.  Yes they have trees so git knows that filex is in directory y because the blob is linked in that tree.  

Let's look at an example.

I create a file in pets folder called currentPet.txt.  It contains 
```
cat
```
in the file.  I commit this file which creates a blob( let's call it bCat).  A tree is created( let's call it TPets. and it knows it is Pets).  TPets points to bCat.  The commit called cCat points to tCat and has other data.  

Work is done.  You now move currentPet.txt from Pets folder and move it to Household folder.  You commit this which creates cCat1 with THouseHold with bCat.  Notice I didn't change the name of the blob that contains the contents of the file.  That's because git reused the same blob.  A different tree is pointing to the blob but it is still using the same blob.  

Git doesn't track filepathX has contentx.  It just builds a tree of the file paths that are changed.  The cool thing is that trees can point to other trees and blobs.  So you have a repo with 100 files in it.  You commit with one file changed.  So the new root tree (the one the commit points to) will point to a hierarchy of trees representing the structure of the folders/files currently committed.  The blob that you just committed will be pointed to by a new tree which will be pointed to by another tree and so on till you get to the root tree that the commit points to.  

So this is why renames in git are tricky.  Context of what head is pointing to, whether you want renames or not and other factors play a role in getting history.

**Comment by vbjay** [comment]
We also have settings that affect how we follow renames.  
![image](https://user-images.githubusercontent.com/3629489/51296595-d9491d80-19ea-11e9-96bb-52b7fb5816b1.png)

**Comment by vbjay** [comment]
We really should be using --follow.  We could allow the user to adjust similarity index percentages.  Let them pick 90% for 90% and above match.

**Comment by vbjay** [comment]
--Find-renames doesn't do what you think.  We need --follow.  https://git-scm.com/docs/git-log#git-log---find-renamesltngt

**Comment by RussKie** [comment]
Jay you seem to know a hell of a lot on the subject. Are you able to put
together a PR of some kind?

**Comment by vbjay** [comment]
I'll look at it. Sure.

**Comment by vbjay** [comment]
Untill my pull request is pulled in, 6005. Canary will not run.

**Comment by vbjay** [comment]
I have it showing the commits.  
![image](https://user-images.githubusercontent.com/3629489/51427652-69908980-1bc8-11e9-9c45-a25ff88dfa86.png)


It just doesn't show anything in the tabs when the file name was changed 
![image](https://user-images.githubusercontent.com/3629489/51427684-b07e7f00-1bc8-11e9-8045-9eed1d76bd9e.png)

**Comment by vbjay** [comment]
GOT IT!  

- Diffs
![image](https://user-images.githubusercontent.com/3629489/51429129-85e8f200-1bd9-11e9-8b8d-945c68093762.png)
![image](https://user-images.githubusercontent.com/3629489/51429136-9bf6b280-1bd9-11e9-9c96-8b53e19f5ec8.png)

The last commit shows nothing so still figuring that out.

**Comment by gerhardol** [comment]
> Is that download stable enough for "productive" use in my daily work or should I expect it to crash/be broken for vital functionality?

@fschmied vbjay has submitted #6132 for this, but it is still work-in-progress. You cannot run the build from the PR until #5990 is merged though unless you build yourself.
Most developers uses private builds from master (with patches for new features) for their day to day work so the canary builds are not unusable. We will want you test the PR when it is ready.

**Comment by CharonX2003** [comment]
Any news on this issue? Is there a branch/build I can test out? We are in the process of reorganizing some of our project files and not having the ability to trace back commits/changes through renames is hurting us more and more...

**Comment by vbjay** [comment]
Rl has taken my time. I plan on continuing this evening.

**Comment by vbjay** [comment]
I have been exhausted from my RL job and personal commitments.  I am winding down from some heavy workloads.  I can get back into this.  

I need to finish up correct git log reading and then finish implementing --follow.

**Comment by CharonX2003** [comment]
Is there any news on this issue? Is it fixed in Version 3.1 or is there any (unstable) branch release I can try out?

**Comment by crypto-rsa** [comment]
I wonder whether this is in any way related to #6458.

**Comment by fschmied** [comment]
Can I do anything to help get this done? It's really slowing me down that GitExtension can no longer follow renames in the history.

**Comment by RussKie** [comment]
If you have capacity, you're welcome to take the lead on the issue.
We would do our best to help, if required.

**Comment by fschmied** [comment]
@vbjay What's the status of your fix?

**Comment by vbjay** [comment]
Sorry. My month-end system just went to production. I haven't had time. As
soon as it has been accepted I will have time. I haven't had personal Dev
time in months. I want to get back but between personal stuff and the
month-end system development.  It's taken my time and energy.

**Comment by fschmied** [comment]
@vbjay No problem, just tell me if/how I can help.

**Comment by gbburkhardt** [comment]
I agree that "Follow Renames" is broken in - Git Extensions 3.0.2.5232
- Build 23b6f51905006ccdde8701591df706284b4155dc
- Git 2.21.0.windows.1
- Microsoft Windows NT 10.0.14393.0
- .NET Framework 4.7.2117.0
- DPI 96dpi (no scaling)

In the display below, only one commit is shown

![file-history](https://user-images.githubusercontent.com/5366676/57455833-11929700-723a-11e9-90ab-f4ef9393c711.png)

But, 'git log --follow' finds two commits:
$ git log --follow taskSetup.m
commit a8f765680b59e0f6db1444658990fa12083b5774 (HEAD -> junk)
Date:   Thu May 9 08:53:52 2019 -0400

    file move

commit b0e46210044f24b2517af1bd5216644206b52a5d (tag: Orig_CC_Import)
Date:   Wed Mar 7 12:08:18 2018 -0500

    Add/Update/remove files to match current CC tools (01-10-2018) /main/Latest ... /main/Latest config in ClearCase.

**Comment by fschmied** [comment]
Okay, I've looked into this a bit and tried to understand the problem and why @vbjay had to do so much work to solve it. I'd like to note it down here for reference.

- A refactoring in https://github.com/gitextensions/gitextensions/commit/a306eb6088b5af56d166f316f4025ca11a23ccef by mistake replaced the `--follow` option with `--format`. This had two effects: the "follow renames" feature was broken and the `git log` command complained about an unrecognized argument. https://github.com/gitextensions/gitextensions/issues/5944#issuecomment-449125777
- https://github.com/gitextensions/gitextensions/commit/00b8daa34e9d15fcb99f632900fd4fc6c9b9dcb1 then removed the false `--format`. This fixed the invalid argument, but did not fix the "follow renames" feature. https://github.com/gitextensions/gitextensions/issues/5944#issuecomment-449128640
- @drewnoakes noticed that the "--follow" was missing, but was reluctant to add it back due to a problem with `--follow` a few years ago (#2363). https://github.com/gitextensions/gitextensions/issues/5944#issuecomment-449159166
- @vbjay looked into the issue and apparently found that just re-adding `--follow` didn't work because while the revisions list contains the right commits, the Diff etc. tabs don't show anything. https://github.com/gitextensions/gitextensions/issues/5944#issuecomment-455782566
- The reason for this is another refactoring, that of `RevisionReader` (formerly `RevisionGraph`). Starting with https://github.com/gitextensions/gitextensions/commit/1f6b0cacc8bc78be076b4e706e5aa030b6d36ea0, the `GitRevision.Name` property is no longer parsed, resulting in the revision items not knowing the file name - which is a problem with renames/moves.
- @vbjay therefore improved the revision log parsing to again include the file names, see PR #6132. This works in principle, but still needs some reliability work (https://github.com/gitextensions/gitextensions/pull/6132/files#r250354949).

So, I now think I fully understand the problem and the pull request #6132.

@vbjay: I will not steal this work from you, but if you feel you're still swamped and want me to take over, I could do so.

**Comment by vbjay** [comment]
👍  for the summary.  I do need to get teh mothballs removed.  My month end project went to prod recently and is in pilot.  😁   So I think I have some time to get back into GE dev.  I really want to finish this.  I am glad you see why it was so much work.

**Comment by vbjay** [comment]
Maybe soon we could work on the begin...end of entries parsing together. Making that work reliably is the first piece.

**Comment by pmiossec** [comment]
> So I think I have some time to get back into GE dev.

👍👍

> I really want to finish this.

So do I! 😁

**Comment by RussKie** [comment]
Top job @fschmied in getting the summary!
Thank you

**Comment by fschmied** [comment]
To help me with my day-to-day work, I made two very quick fixes to restore rename following in the file history, effectively restoring the behavior to the way it worked in GitExt v2.5x. If anyone is interested, they can create a private build of it from https://github.com/fschmied/gitextensions/commit/7b04518e590a14475c9d258453f3350fdf6446ee (based on current master) or https://github.com/fschmied/gitextensions/commit/01158710ef466d6ce57b6a18128ae426b6fcde4f (based on release/3.1).

This does not contain the extensive refactorings from #6132, and I also didn't add any automated tests, so I'm _not_ creating a pull request. This is purely to help me (and maybe others) out until #6132 is done.

**Comment by CharonX2003** [comment]
@fschmied would it maybe possible to obtain a Binary for those fixed branches somehow? I unfortunately do not have VS 2017 and thus find myself unable to build the the project :(

**Comment by mstv** [comment]
You could open a temporary PR and add "WIP" to its title. AppVeyor will build binaries.

**Comment by fschmied** [comment]
I've uplodaed the build of the branch based on release 3.1 I'm currently using to here https://gist.github.com/fschmied/cf6d359b8efba5da794bc034c30a7ee6#file-readme-md.

> You could open a temporary PR and add "WIP" to its title. AppVeyor will build binaries.

Good idea, but I guess the real PR will be done soon anyway.

**Comment by RussKie** [comment]
We need to nudge Jay 😄

**Comment by vbjay** [comment]
Lol. I know. I'm devoting a huge portion of my weekend to getting this
done.

**Comment by Hottemax** [comment]
Works for me again (renamed files are correctly followed)  since the current version released today [3.1.1.6049], although it is not mentioned in the Changelog.md

Thank you!

**Comment by RussKie** [comment]
Great to hear, though I don't think we've done anything in this area... 😕 
Unless, of course, Jay snuck in his fix while I was asleep.

**Comment by forger** [comment]
Nope, still broken for me in 3.1.1.6049 :(

**Comment by Hottemax** [comment]
> Nope, still broken for me in 3.1.1.6049 :(

Do you have `[log] follow = true` in your .gitconfig? Otherwise it does not work ...

**Comment by fschmied** [comment]
@Hottemax Nice idea, but the point of this issue is that GitExtensions should follow renames (if the feature is turned on within GitExtensions) independently from whether `[log] follow = true` is configured or not.

If you try to work around the bug using the git config, you'll likely see the behavior described in https://github.com/gitextensions/gitextensions/issues/5944#issuecomment-455782566 - the history list view will follow the renames, but the tabs below won't know what to do with the corresponding commits.

**Comment by Hottemax** [comment]
@fschmied You are right of course, it should work independent of your gitconfig.
I was merely remarking that it works in the current version when this option is set, as opposed to before.

**Comment by fschmied** [comment]
@Hottemax Can you also click on all the history entries (before the move) and see something in the tabs below (Commit, Diff, ...)?

**Comment by Hottemax** [comment]
> @Hottemax Can you also click on all the history entries (before the move) and see something in the tabs below (Commit, Diff, ...)?

Yes.

**Comment by fschmied** [comment]
@Hottemax I finally found the time to test the workaround, but unfortunately, it works only half-way for me. While `[log] follow=true` will cause all the commits to be shown for a file (yes!), three of the tabs below don't work for commits prior to the rename.

![grafik](https://user-images.githubusercontent.com/139564/59562297-45c45900-902b-11e9-8c71-841bd4045f7d.png)

**Comment by fschmied** [comment]
I've followed @mstv's [advice](https://github.com/gitextensions/gitextensions/issues/5944#issuecomment-496663490) and have created a WIP PR for my quickfix to get a new build based on 3.1.1: #6807.

The latest build is here:  https://ci.appveyor.com/project/gitextensions/gitextensions/builds/25312244/artifacts.

**Comment by gbburkhardt** [comment]
Excellent.

**Comment by jrfnl** [comment]
Looks like this is still not completely fixed.

See: https://github.com/gitextensions/gitextensions/pull/6807#issuecomment-518429916

Let me know if you need more info.

**Comment by gerhardol** [comment]
@jfrnl
The fix in master should be the same functionality as in 2.x
Renames are not finished - #6132 is a wip for a more complete solution
See @vbjay explanation and links to why renames are not trivial, it is not tracked in Git and has to be calculated.

**Comment by IanKemp** [comment]
When can we expect a 3.2 release including the fix from #6879? It's not perfect but it's far better than no history at all...

**Comment by pmiossec** [comment]
@IanKemp We hope to release a version in a not so long time even if we release when we are confident enough.

But if you can't wait, you could try the development version that is working quite well: https://github.com/gitextensions/gitextensions#downloads

**Comment by IanKemp** [comment]
> @IanKemp We hope to release a version in a not so long time even if we release when we are confident enough.
> 
> But if you can't wait, you could try the development version that is working quite well: https://github.com/gitextensions/gitextensions#downloads

Thanks for that - is it safe to download the latest `Portable.zip` and overwrite the contents of my current Git Extensions install? In other words, would that break my install? (Not talking about potential breakage of config files etc. - that is an accepted and understood risk by me.)

**Comment by RussKie** [comment]
Yes, it is safe. Just update the config file and unmark the app as portable.

**Comment by fschmied** [comment]
Is this fix in 3.2 and 3.2.1? (3.2 doesn't have any release notes, and the notes for 3.2.1 don't seem to mention this issue.)

**Comment by pmiossec** [comment]
Yes.

**Comment by forger** [comment]
Works like a charm again <3 This issue has been a major drag for me over the last months. Thanks to all involved in solving it!

**Comment by Kryptos-FR** [comment]
Still not working for me. Or do we really need to have `[log] follow = true`/ If so GitExtensions should set it for you if the feature is enabled.
